### DISCUSSION TITLE
How to Verify certificate is installed correctly or not?

### DISCUSSION THREAD

**Discussion by YogarajRethinasamy** [International-Data-Spaces-Association/DataspaceConnector#89] [Q&A] [answered]
Hi,

I have got the certificate from fraunhofer, as per the documentation **[https://international-data-spaces-association.github.io/DataspaceConnector/Deployment/Configuration#step-2-ids-certificate]** i have updated everything.

But i dont know how to verify my certificate is installed correctly or not

Kindly help me out on this

**Comment by ronjaquensel** [comment, marked as answer]
Hi,

you can verify this by sending a description request to your own connector (*/api/ids/description* with recipient *[your-connector-url]/api/ids/data*).

If the *connectorDeployMode* is set to productive deployment and no valid IDS certificate is configured, the message sending will fail and you will see the following logs:

```
ERROR - PRODUCTIVE_DEPLOYMENT: No IDS-Message sent! No DAT could be acquired from DAPS! [reason=(Mandatory required information of the connector certificate is missing (AKI/SKI)! Needed to determine the fingerprint of the connector for the DAPS! Possible reason: Are you using a valid connector certificate issued by the DAPS?)]
```

If the *connectorDeployMode* is set to productive deployment and a valid IDS certificate is correctly configured, you will receive the connector's self-description and see the following logs:

```
INFO - Successfully received DAT from DAPS.
[...]
INFO - Successfully verified DAT claims.
```

**Reply by YogarajRethinasamy** [reply]
Thanks, I have got the below error, can you guide me what are the thing i need to do inorder to fix this


2021-09-17T07:27:01,823 [https-jsse-nio-6443-exec-9] INFO - Init duration for springdoc-openapi is: 2548 ms
2021-09-17T07:27:15,945 [https-jsse-nio-6443-exec-1] WARN - TEST_DEPLOYMENT: IDS-Message is sent without a valid DAT, will not be send in PRODUCTIVE_DEPLOYMENT, reason: Mandatorily required information of the connector certificate is missing (AKI/SKI)!
2021-09-17T07:27:16,129 [https-jsse-nio-6443-exec-1] ERROR - JWT strings must contain exactly 2 period characters. Found: 0
2021-09-17T07:27:16,129 [https-jsse-nio-6443-exec-1] WARN - Could not parse jwt!
2021-09-17T07:27:16,129 [https-jsse-nio-6443-exec-1] WARN - TEST_DEPLOYMENT: IDS-Message is sent without a valid DAT, will not be send in PRODUCTIVE_DEPLOYMENT, reason: Mandatorily required information of the connector certificate is missing (AKI/SKI)!
2021-09-17T07:27:16,234 [https-jsse-nio-6443-exec-1] INFO - URL is valid: null

**Reply by ghost** [reply]
Log-Error-Message can be ignored, was removed with IDS-Messaging-Services v4.0.0 (included in Dataspace Connector from v6.1.3).

**Reply by ronjaquensel** [reply]
Did you set the *connectorDeployMode* in the `config.json` to *idsc:PRODUCTIVE_DEPLOYMENT*? The warn-logs indicate that the connector is running in test mode rather than productive mode. If *idsc:PRODUCTIVE_DEPLOYMENT* is set in your `config.json`, the connector might be loading older configuration from the database. In that case, could you start the connector with a clean database and try again?

**Reply by YogarajRethinasamy** [reply]
Yes i have changed that to   PRODUCTIVE_DEPLOYMENT in json but still it showing the error


"ids:connectorDeployMode" : {
    "@id" : "idsc:PRODUCTIVE_DEPLOYMENT"

**Reply by YogarajRethinasamy** [reply]
currently my certificate is xxxx.p12 do i need to give the same xxxx.p12 name or i need to change the name of the given certificate file to default one like keystore-localhost.p12 

how to use the given certificate, FYI: i have already placed  all certificate related it in conf folder as well

**Reply by ronjaquensel** [reply]
Your keystore file can have any name, as long as the same name is used in the `config.json`. So if your file is called *xxx.p12* that's fine as long as you modify the `config.json` to

```
"ids:keyStore" : {
    "@id" : "file:///conf/xxx.p12"
  }
```

Did you clear the database before starting the connector?

**Reply by YogarajRethinasamy** [reply]
currently we don't have any data in db, Also can you guide me how to clear the db?

Also am getting the warning message as well while starting the application

2021-09-17T08:30:51,371 [main] WARN - Trustmanager is trusting all Certificates in TEST_DEPLOYMENT mode, you should not use this in production!

**Reply by YogarajRethinasamy** [reply]
still the same error presists

 WARN - TEST_DEPLOYMENT: IDS-Message is sent without a valid DAT, will not be send in PRODUCTIVE_DEPLOYMENT, reason: Mandatorily required information of the connector certificate is missing (AKI/SKI)!
2021-09-17T08:32:29,454 [https-jsse-nio-6443-exec-6] ERROR - JWT strings must contain exactly 2 period characters. Found: 0
2021-09-17T08:32:29,455 [https-jsse-nio-6443-exec-6] WARN - Could not parse jwt!
2021-09-17T08:32:29,455 [https-jsse-nio-6443-exec-6] WARN - TEST_DEPLOYMENT: IDS-Message is sent without a valid DAT, will not be send in PRODUCTIVE_DEPLOYMENT, reason: Mandatorily required information of the connector certificate is missing (AKI/SKI)!

**Reply by ronjaquensel** [reply]
Once the connector is started, it will persist its configuration in the database and will reload it for every subsequent start. Therefore, the config.json does not have any effect, if there is already a configuration in the database. How to clear the database depends on the database you're using.

**Reply by YogarajRethinasamy** [reply]
FYI: We are not using any DB, i have just cloned the git repo & changed the certificates that are provided by franhoufer.

**Reply by ghost** [reply]
Did you try a maven clean and maven package, after the config and the certificates have been adjusted (in case you should start the connector as jar)?

**Reply by YogarajRethinasamy** [reply]
Just now i tried to do  clean maven, now build failure

[root@ip-10-61-20-115 DataspaceConnector]# ./mvnw clean package
[INFO] Scanning for projects...
[INFO]
[INFO] -----------------------< io:dataspaceconnector >------------------------
[INFO] Building Dataspace Connector 6.1.0
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ dataspaceconnector ---
[INFO] Deleting /root/DataspaceConnector/target
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-maven) @ dataspaceconnector ---
[INFO]
[INFO] --- maven-resources-plugin:3.2.0:resources (default-resources) @ dataspaceconnector ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 12 resources
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.866 s
[INFO] Finished at: 2021-09-17T11:19:25Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-resources-plugin:3.2.0:resources (default-resources) on project dataspaceconnector: Input length = 1 -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:

**Reply by ronjaquensel** [reply]
This error might occur due to invalid characters in the files located in the *resources* folder. What changes did you make after cloning the repository?

**Reply by YogarajRethinasamy** [reply]
These are the changes that i made after receiving certificates from fraunhour, can you kindly guide me is there anything else that i need to change apart from these 
 
 config.json
 ========
 "ids:connectorDeployMode" : {
    "@id" : "idsc:**PRODUCTIVE_DEPLOYMENT**"
  },
  "ids:connectorDescription" : {
    "@type" : "ids:BaseConnector",
    "@id" : "https://**xxxx.ids.com**",
	  
"ids:hasDefaultEndpoint" : {
      "@type" : "ids:ConnectorEndpoint",
      "@id" : "https://w3id.org/idsa/autogen/connectorEndpoint/e5e2ab04-633a-44b9-87d9-a097ae6da3cf",
      "ids:accessURL" : {
        "@id" : "https://**xxx.ids.com**/api/ids/data"
      }
  "ids:connectorStatus" : {
    "@id" : "idsc:**CONNECTOR_ONLINE**"
  },
  "ids:keyStore" : {
    "@id" : "file:///conf/**xxx.ids.com.p12**"
  }

appilication.propereities
=================
## TLS
server.ssl.enabled=true
server.ssl.key-store-type=PKCS12
server.ssl.key-store=classpath:conf/**xxx.ids.com.p12**
server.ssl.key-store-password=password
server.ssl.key-alias=1
#security.require-ssl=true


## IDSCP
application.http.base-url=https://**xxx.ids.com**
idscp2.enabled=false
idscp2.keystore=./src/main/resources/conf/**xxx.ids.com.p12**
idscp2.truststore=./src/main/resources/conf/truststore.p12

Error
====

[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ContractAgreementHandlerTest.handleMessage_validAgreement_returnMessageProcessedNotificationMessageAndLogToCH:248 ClassCast
[ERROR]   ContractRequestHandlerTest.checkContractRequest_validRequest_returnOk:493 ClassCast
[ERROR]   DescriptionRequestHandlerTest.handleMessage_validResourceDescriptionMsgKnownId_returnResourceDescription:176 ClassCast
[ERROR]   DescriptionRequestHandlerTest.handleMessage_validSelfDescriptionMsg_returnSelfDescription:114 ClassCast
[INFO]
[ERROR] Tests run: 1341, Failures: 0, Errors: 4, Skipped: 1
[INFO]
